File: Ejercicio 9/Ejercicio 9.cs
using System;
using System.Collections.Generic;

public class MergeSort
{
    // Método para combinar dos subarrays ordenados
    private static void Merge(int[] array, int left, int middle, int right)
    {
        // Tamaños de los subarrays temporales
        int n1 = middle - left + 1;
        int n2 = right - middle;
        
        // Arrays temporales
        int[] leftArray = new int[n1];
        int[] rightArray = new int[n2];
        
        // Copia datos a los arrays temporales
        for (int i = 0; i < n1; i++)
        {
            leftArray[i] = array[left + i];
        }
        for (int j = 0; j < n2; j++)
        {
            rightArray[j] = array[middle + 1 + j];
        }
        
        // Combina los arrays temporales
        int i_idx = 0, j_idx = 0;
        int k = left;
        
        while (i_idx < n1 && j_idx < n2)
        {
            if (leftArray[i_idx] <= rightArray[j_idx])
            {
                array[k] = leftArray[i_idx];
                i_idx++;
            }
            else
            {
                array[k] = rightArray[j_idx];
                j_idx++;
            }
            k++;
        }
        
        // Copia los elementos restantes de leftArray (si hay)
        while (i_idx < n1)
        {
            array[k] = leftArray[i_idx];
            i_idx++;
            k++;
        }
        
        // Copia los elementos restantes de rightArray (si hay)
        while (j_idx < n2)
        {
            array[k] = rightArray[j_idx];
            j_idx++;
            k++;
        }
    }
    
    // Método principal que implementa Merge Sort
    public static void MergeSortArray(int[] array, int left, int right)
    {
        if (left < right)
        {
            // Encuentra el punto medio
            int middle = (left + right) / 2;
            
            // Ordena la primera y segunda mitad
            MergeSortArray(array, left, middle);
            MergeSortArray(array, middle + 1, right);
            
            // Combina las mitades ordenadas
            Merge(array, left, middle, right);
        }
    }
    
    // Método para generar números aleatorios
    public static int[] GenerarArrayAleatorio(int tamaño, int maximo)
    {
        int[] array = new int[tamaño];
        Random random = new Random();
        
        for (int i = 0; i < tamaño; i++)
        {
            array[i] = random.Next(1, maximo + 1); // Números entre 1 y maximo
        }
        
        return array;
    }
    
    // Función para imprimir un array
    public static void ImprimirArray(int[] array)
    {
        Console.Write("[");
        for (int i = 0; i < array.Length; i++)
        {
            Console.Write(array[i]);
            if (i < array.Length - 1)
            {
                Console.Write(", ");
            }
        }
        Console.Write("]");
    }
    
    // Método adicional para demostrar el proceso paso a paso
    public static void DemostrarMergeSort()
    {
        int[] demo = {38, 27, 43, 3, 9, 82, 10};
        Console.WriteLine("Ejemplo paso a paso con array pequeño:");
        Console.Write("Array inicial: ");
        ImprimirArray(demo);
        Console.WriteLine();
        
        MergeSortArray(demo, 0, demo.Length - 1);
        
        Console.Write("Array final: ");
        ImprimirArray(demo);
        Console.WriteLine();
    }
    
    // Método principal
    public static void Main(string[] args)
    {
        // Generar 15 números aleatorios entre 1 y 100
        int[] numeros = GenerarArrayAleatorio(15, 100);
        
        Console.WriteLine("=== ORDENAMIENTO MERGE SORT ===");
        Console.WriteLine("Array original:");
        ImprimirArray(numeros);
        Console.WriteLine();
        
        // Aplicar Merge Sort
        MergeSortArray(numeros, 0, numeros.Length - 1);
        
        Console.WriteLine("\nArray ordenado:");
        ImprimirArray(numeros);
        Console.WriteLine();
        
        // Mostrar paso a paso (opcional)
        Console.WriteLine("\n=== DEMOSTRACION PASO A PASO ===");
        DemostrarMergeSort();
    }
}

File: Ejercicio 9/Ejercicio 9.py
import random

def merge_sort(array, left, right):
    """
    Método principal que implementa Merge Sort
    """
    if left < right:
        # Encuentra el punto medio
        middle = (left + right) // 2
        
        # Ordena la primera y segunda mitad
        merge_sort(array, left, middle)
        merge_sort(array, middle + 1, right)
        
        # Combina las mitades ordenadas
        merge(array, left, middle, right)

def merge(array, left, middle, right):
    """
    Método para combinar dos subarrays ordenados
    """
    # Tamaños de los subarrays temporales
    n1 = middle - left + 1
    n2 = right - middle
    
    # Arrays temporales
    left_array = [0] * n1
    right_array = [0] * n2
    
    # Copia datos a los arrays temporales
    for i in range(n1):
        left_array[i] = array[left + i]
    for j in range(n2):
        right_array[j] = array[middle + 1 + j]
    
    # Combina los arrays temporales
    i = j = 0
    k = left
    
    while i < n1 and j < n2:
        if left_array[i] <= right_array[j]:
            array[k] = left_array[i]
            i += 1
        else:
            array[k] = right_array[j]
            j += 1
        k += 1
    
    # Copia los elementos restantes de left_array (si hay)
    while i < n1:
        array[k] = left_array[i]
        i += 1
        k += 1
    
    # Copia los elementos restantes de right_array (si hay)
    while j < n2:
        array[k] = right_array[j]
        j += 1
        k += 1

def generar_array_aleatorio(tamaño, maximo):
    """
    Método para generar números aleatorios
    """
    array = [0] * tamaño
    
    for i in range(tamaño):
        array[i] = random.randint(1, maximo)  # Números entre 1 y maximo
    
    return array

def demostrar_merge_sort():
    """
    Método adicional para demostrar el proceso paso a paso
    """
    demo = [38, 27, 43, 3, 9, 82, 10]
    print("Ejemplo paso a paso con array pequeño:")
    print(f"Array inicial: {demo}")
    
    merge_sort(demo, 0, len(demo) - 1)
    
    print(f"Array final: {demo}")

def main():
    """
    Método principal
    """
    # Generar 15 números aleatorios entre 1 y 100
    numeros = generar_array_aleatorio(15, 100)
    
    print("=== ORDENAMIENTO MERGE SORT ===")
    print("Array original:")
    print(numeros)
    
    # Aplicar Merge Sort
    merge_sort(numeros, 0, len(numeros) - 1)
    
    print("\nArray ordenado:")
    print(numeros)
    
    # Mostrar paso a paso (opcional)
    print("\n=== DEMOSTRACION PASO A PASO ===")
    demostrar_merge_sort()

if __name__ == "__main__":
    main()